### DISCUSSION TITLE
Connect to topic from javascript?

### DISCUSSION THREAD

**Discussion by philmday** [permitio/fastapi_websocket_pubsub#18] [Q&A]
Is it possible to get a javascript frontend to subscribe to a topic?  I want to output data in real-time to a graph from my multiple data crunching python processes.

**Comment by asafc** [comment]
Hi @philmday :)
Currently fastapi_websocket_pubsub is built for RPC/pubsub connections between backend services.
Frontend syncing **is not** one of our use cases.

You could technically implement a javascript client connecting directly to the python pubsub server, but i would suggest a different (and simpler) approach.

You have 2 main entities in my suggested solution:
* **frontend** - javascript client (e.g: react)
* **backend** - backend server in python (e.g: with fastapi)
  * the backend server will run a pubsub client in the background

Solution outline:
* In your **backend** (the one serving the frontend js clients), run a pubsub client in an async task.
* Your pubsub client should subscribe to ALL_TOPICS.
* In your ALL_TOPICS callback, you can then pass / not pass the data to the connected **frontend** clients depending on the topic (whether or not the frontend clients "subscribed" to it.
* You would need to implement some subscribe/unsubscribe logic in your **backend** and manage the connections.
  * e.g: implement a `/subscribe` and `/unsubscribe` HTTP endpoints in your backend the javascript can call with a topic.
* You should pass your connection manager object to the pubsub client **callback method** so it knows where it should pipeline messages.
* In your specific use case i am not sure websockets between your **frontend** and your **backend** is the best solution since you don't need bidirectional connectivity. Take a look at SSE (server sent events).

Hope it helps :)

**Reply by orweis** [reply]
I agree with @asafc; but I would add that if you want, you can rather easily implement a JS version of the client.
The place to start is with FastAPI-Websocket-RPC's RPC channel, and then add the PubSub on top. 

If you (or anyone else) are interested in doing so I'd be happy to walk you through the code, so you can contribute this to the project. Otherwise we might add this ourselves in the future (no immediate plans yet).

**Reply by iloveitaly** [reply]
> Frontend syncing **is not** one of our use cases.

Is there a good alternative library that supports [actioncable-like](https://guides.rubyonrails.org/action_cable_overview.html) functionality in FastAPI? This is the closest thing I'm able to find.

**Reply by orweis** [reply]
> Is there a good alternative library that supports [actioncable-like](https://guides.rubyonrails.org/action_cable_overview.html) functionality in FastAPI? This is the closest thing I'm able to find.


There are some example JS snippets for this people shared - check out https://github.com/permitio/fastapi_websocket_pubsub/discussions/42

**Reply by iloveitaly** [reply]
Got it, but no combined JS + py libraries to work off of at this point?

**Reply by orweis** [reply]
Afraid so.
If you want to give it a shot in creating those yourself for open-source, I'd be happy to provide guidance, and also send you some Permit.io swag as extra incentive 😅😇
